flink-runtime/src/test/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPoolTest.java

```diff
@@ -0,0 +1,286 @@
+            }
+            assertEquals(bufferPool.getNumTotalBuffers(), bufferPool.getAvailableBuffers());
+            bufferPool.destroy();
+        }
+    }
+
+    @Test
+    public void testMultipleThreadRequestAndRecycle() throws Exception {
+        int numRequestThreads = 10;
+        AtomicReference<Throwable> exception = new AtomicReference<>();
+        BatchShuffleReadBufferPool bufferPool = createBufferPool();
+
+        try {
+            Thread[] requestThreads = new Thread[numRequestThreads];
+            for (int i = 0; i < numRequestThreads; ++i) {
+                requestThreads[i] =
+                        new Thread(
+                                () -> {
+                                    try {
+                                        Object owner = new Object();
+                                        for (int j = 0; j < 100; ++j) {
+                                            List<MemorySegment> buffers =
+                                                    bufferPool.requestBuffers(owner);
+                                            Thread.sleep(10);
+                                            if (j % 2 == 0) {
+                                                bufferPool.recycle(buffers, owner);
+                                            } else {
+                                                for (MemorySegment segment : buffers) {
+                                                    bufferPool.recycle(segment, owner);
+                                                }
+                                            }
+                                        }
+                                    } catch (Throwable throwable) {
+                                        exception.set(throwable);
+                                    }
+                                });
+                requestThreads[i].start();
+            }
+
+            for (Thread requestThread : requestThreads) {
+                requestThread.join();
+            }
+
+            assertNull(exception.get());
+            assertEquals(bufferPool.getNumTotalBuffers(), bufferPool.getAvailableBuffers());
+        } finally {
+            bufferPool.destroy();
+        }
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void testDestroy() throws Exception {
+        BatchShuffleReadBufferPool bufferPool = createBufferPool();
+        bufferPool.initialize();
+
+        assertFalse(bufferPool.isDestroyed());
+        assertEquals(bufferPool.getNumTotalBuffers(), bufferPool.getAvailableBuffers());
+
+        List<MemorySegment> buffers = bufferPool.requestBuffers(this);
+        assertEquals(
+                bufferPool.getNumTotalBuffers() - buffers.size(), bufferPool.getAvailableBuffers());
+
+        bufferPool.destroy();
+        assertTrue(bufferPool.isDestroyed());
+        assertEquals(0, bufferPool.getAvailableBuffers());
+
+        bufferPool.recycle(buffers, this);
```

Comment:
maybe add a separate test `testRecycleAfterPoolDestroyed` to test that recycled buffers will be directly discarded if the pool is destroyed?